src/js/reducers/conversation-reducer.js

```diff
@@ -36,22 +34,75 @@ const addMessage = (messages, message) => {
     return sortMessages([...messages, message]);
 };
 
+
+const matchMessage = (message, queryProps) => Object.keys(queryProps).every((key) => message[key] === queryProps[key]);
+
+const replaceMessage = (messages, query, newMessage) => {
+    const existingMessage = messages.find((message) => matchMessage(message, query));
+    if (!existingMessage) {
+        return messages;
+    }
+
+    const index = messages.indexOf(existingMessage);
+    return [...messages.slice(0, index), newMessage, ...messages.slice(index + 1)];
+};
+
+const isEqual = (messageA, messageB) => {
+    if (messageA._id && messageB._id && messageA._id === messageB._id) {
+        return true;
+    }
+
+    if (!messageA._id || !messageB._id) {
+        if (messageA.role === messageB.role) {
+            if (messageA.text && messageB.text && messageA.text === messageB.text) {
+                return true;
+            }
+
+            if (messageA.mediaType === messageB.mediaType && messageA.mediaUrl === messageB.mediaUrl) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+};
+
+const mergeMessages = (messagesA, messagesB) => {
+    // concat will make a union out of both arrays
+    return removeDuplicates(messagesA.concat(messagesB));
+};
+
+const removeDuplicates = (messages) => {
+    // reduce will return a new array, the function used in the reduction strips out duplicates
+    return messages.reduce((filteredMessages, nextMessage) => {
+        const message = filteredMessages.find((m) => isEqual(m, nextMessage));
+        if (message) {
+            return filteredMessages;
+        }
+        return [...filteredMessages, nextMessage];
+    }, []);
+};
+
 export function ConversationReducer(state = INITIAL_STATE, action) {
     switch (action.type) {
         case RESET:
         case ConversationActions.RESET_CONVERSATION:
             return Object.assign({}, INITIAL_STATE);
         case ConversationActions.SET_CONVERSATION:
             return Object.assign({}, action.conversation, {
-                messages: sortMessages(action.conversation.messages)
+                messages: sortMessages(mergeMessages(state.messages, action.conversation.messages))
             });
         case ConversationActions.ADD_MESSAGE:
             return Object.assign({}, state, {
                 messages: addMessage(state.messages, action.message)
             });
+        case ConversationActions.REPLACE_MESSAGE:
+            return Object.assign({}, state, {
+                messages: removeDuplicates(replaceMessage(state.messages, action.queryProps, action.message))
```

Comment:
Image upload works in 3 steps:
- ADD_MESSAGE of the "sending" image [1]
- ADD_MESSAGE of the "received" image [2]
- REPLACE_MESSAGE of the "sending" by a different "received" [3]
This video makes it more clear: https://www.dropbox.com/s/vw6lcx86ddj0ri5/uplading_image.mov?dl=0

I added the removeDuplicates because I would end up with two copies of the "received" image.

Although, it might be worthwhile to re-visit the replaceMessage logic.